diff --git a/features.export.inc b/features.export.inc
index 2f32c95..4db8f8a 100644
--- a/features.export.inc
+++ b/features.export.inc
@@ -582,18 +582,6 @@ function features_var_export($var, $prefix = '', $init = TRUE, $count = 0) {
   else if (is_int($var)) {
     $output = intval($var);
   }
-  else if (is_numeric($var)) {
-    $floatval = floatval($var);
-    if (is_string($var) && ((string) $floatval !== $var)) {
-      // Do not convert a string to a number if the string
-      // representation of that number is not identical to the
-      // original value.
-      $output = var_export($var, TRUE);
-    }
-    else {
-      $output = $floatval;
-    }
-  }
   else if (is_string($var) && strpos($var, "\n") !== FALSE) {
     // Replace line breaks in strings with a token for replacement
     // at the very end. This protects whitespace in strings from
