diff --git a/redirect.module b/redirect.module
index 9f7a256..0bfef8a 100644
--- a/redirect.module
+++ b/redirect.module
@@ -986,7 +986,6 @@ function redirect_redirect($redirect = NULL) {
   }
   else {
     watchdog('redirect', 'Infinite loop stopped.');
-    drupal_set_message('Oops, looks like this request tried to create an infinite loop. We do not allow such things here. We are a professional website!');
     return FALSE;
   }
 
