diff --git a/reply.module b/reply.module
index 359b08c..904e934 100644
--- a/reply.module
+++ b/reply.module
@@ -1674,12 +1674,15 @@ function reply_page_list() {
     array('data' => t('Operations'), 'colspan' => 3)
   );
 
-  $ids = db_select('reply', 'r')
+  $parameters = drupal_get_query_parameters();
+  $query = db_select('reply', 'r')
     ->fields('r', array('id'))
     ->extend('PagerDefault')->limit(20)
-    ->extend('TableSort')->orderByHeader($header)
-    ->execute()
-    ->fetchCol();
+    ->extend('TableSort')->orderByHeader($header);
+  if (isset($parameters['bundle'])) {
+    $query->condition('r.bundle', $_GET['bundle']);
+  }
+  $ids = $query->execute()->fetchCol();
 
   $languages = language_list();
   $loadBundles = reply_load_bundles();
