diff --git a/editors/ckeditor.inc b/editors/ckeditor.inc
index 6ff60d5..c66d173 100644
--- a/editors/ckeditor.inc
+++ b/editors/ckeditor.inc
@@ -193,7 +193,12 @@ function wysiwyg_ckeditor_init($editor) {
   // reliable. The official documentation suggests to solve the issue through
   // the global window variable.
   // @see http://docs.cksource.com/CKEditor_3.x/Developers_Guide/Specifying_the_Editor_Path
-  $library_path = base_path() . $editor['library path'] . '/';
+  $base_path = base_path();
+  if (substr($base_path, 0, 2) == '//'){
+    $scheme = empty($_SERVER['HTTP_X_FORWARDED_PROTO']) ? 'http:' : $_SERVER['HTTP_X_FORWARDED_PROTO'];
+    $base_path = $scheme . $base_path;
+  }
+  $library_path = $base_path . $editor['library path'] . '/';
   return <<<EOL
 window.CKEDITOR_BASEPATH = '$library_path';
 EOL;
